compiler/rustc_lint/src/builtin.rs

```diff
@@ -1839,6 +1839,72 @@ impl EarlyLintPass for IncompleteFeatures {
     }
 }
 
+declare_lint! {
+    pub INCOMPATIBLE_FEATURES,
+    Deny,
+    "Incompatible features that would result in an undefined behavior if present at the same time"
+}
+
+declare_lint_pass!(
+    /// Check for used feature gates in `INCOMPATIBLE_FEATURES` in `librustc_feature/active.rs`.
+    IncompatibleFeatures => [INCOMPATIBLE_FEATURES]
+);
+
+impl EarlyLintPass for IncompatibleFeatures {
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, _: &ast::Crate) {
+        // track which features we have visited from the incompatible list
+        let mut visited = [(false, false); rustc_feature::INCOMPATIBLE_FEATURES.len()];
+
+        let features = cx.sess.features_untracked();
+        let filtered_features: Vec<(&Symbol, &Span)> = features
```

Comment:
this seems somewhat complicated to me :thinking:

It might be cleaner to instead walk `INCOMPATIBLE_FEATURES` once and test `features.enabled` for both.

If we find a match we should then be able to search the feature in `declared_lang_features` + `declared_lib_features` to get the correct span.

Reply:
Yah, I found it complicated too, thanks for the suggestion.